pkg/controller/seed-controller-manager/etcdrestore/etcd_restore_controller.go

```diff
@@ -0,0 +1,267 @@
+
+	incompleteRestorePredicates := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			restore := e.Object.(*kubermaticv1.EtcdRestore)
+			return restore.Status.Phase != kubermaticv1.EtcdRestorePhaseCompleted
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			restore := e.ObjectNew.(*kubermaticv1.EtcdRestore)
+			return restore.Status.Phase != kubermaticv1.EtcdRestorePhaseCompleted
+		},
+	}
+
+	return c.Watch(&source.Kind{Type: &kubermaticv1.EtcdRestore{}}, &handler.EnqueueRequestForObject{}, incompleteRestorePredicates)
+}
+
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	log := r.log.With("request", request)
+	log.Debug("Processing")
+
+	restore := &kubermaticv1.EtcdRestore{}
+	if err := r.Get(ctx, request.NamespacedName, restore); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, types.NamespacedName{Name: restore.Spec.Cluster.Name}, cluster); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	log = r.log.With("cluster", cluster.Name, "restore", restore.Name)
+
+	if cluster.Labels[kubermaticv1.WorkerNameLabelKey] != r.workerName {
+		return reconcile.Result{}, nil
+	}
+
+	result, err := r.reconcile(ctx, log, restore, cluster)
+	if err != nil {
+		log.Errorw("Reconciling failed", zap.Error(err))
+		r.recorder.Event(restore, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+		r.recorder.Eventf(cluster, corev1.EventTypeWarning, "ReconcilingError",
+			"failed to reconcile etcd restore %q: %v", restore.Name, err)
+	}
+
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, restore *kubermaticv1.EtcdRestore, cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	log.Infof("performing etcd restore from backup %v", restore.Spec.BackupName)
+
+	if restore.Status.Phase == kubermaticv1.EtcdRestorePhaseCompleted {
+		return nil, nil
+	}
+
+	if restore.Status.Phase == kubermaticv1.EtcdRestorePhaseStsRebuilding {
+		return r.rebuildEtcdStatefulset(ctx, log, restore, cluster)
+	}
+
+	// pause cluster
+	if err := r.updateCluster(ctx, cluster, func(cluster *kubermaticv1.Cluster) {
+		cluster.Spec.Pause = true
+	}); err != nil {
+		return nil, fmt.Errorf("failed to pause cluster: %v", err)
+	}
+
+	if err := r.updateRestore(ctx, restore, func(restore *kubermaticv1.EtcdRestore) {
+		restore.Status.Phase = kubermaticv1.EtcdRestorePhaseStarted
+		kuberneteshelper.AddFinalizer(restore, RebuildStatefulsetFinalizer)
+	}); err != nil {
+		return nil, fmt.Errorf("failed to add finalizer: %v", err)
+	}
+
+	// delete etcd sts
+	sts := &v1.StatefulSet{}
+	err := r.Get(ctx, types.NamespacedName{Namespace: cluster.Status.NamespaceName, Name: resources.EtcdStatefulSetName}, sts)
```

Comment:
I think we agreed this would be done only when we have a new storage class set.

Reply:
As we talked about in Slack briefly, I've postponed this for now because it wasn't so easy to implement reliably. When deleting and immediately recreating all etcd pods, it happened sometimes that data would "spill" from still-running old pods to already-running, so that after the restore the data in etcd sometimes contained not just data that was in the backup, but also some keys that had been added after the backup was taken. I tried to circumvent this by scaling the sts down to 0 and then up again, but currently the cluster resource controller can't scale the etcd sts up from 0 (look at https://github.com/kubermatic/kubermatic/blob/c764c003af56e1c11090e20d314d5806e7b83645/pkg/resources/etcd/statefulset.go#L353-L384 for the case *set.Spec.Replicas==0 to see this -- actually not sure if this is a bug or not).

I've pushed the "delete all pods" and "scale sts down/up" implementations here
https://github.com/multi-io/kubermatic/blob/etcd-backups-controller-del-sts-pods/pkg/controller/seed-controller-manager/etcdrestore/etcd_restore_controller.go and here
https://github.com/multi-io/kubermatic/blob/etcd-backups-controller-scaledown-sts/pkg/controller/seed-controller-manager/etcdrestore/etcd_restore_controller.go in case we want to revisit this.